src/test/py/bazel/bzlmod/mod_command_test.py

```diff
@@ -239,6 +241,65 @@ def testGraphWithExtensionFilter(self):
         'wrong output in graph query with extension filter specified',
     )
 
+  def testGraphWithFailingExtensions(self):
+    # Force ext2 to fail.
+    with open(self.Path('MODULE.bazel'), 'a+') as f:
+      f.write("ext2.dep(name = 'repo2')\n")
+      f.write("ext2.dep(name = 'fail')\n")
+
+    exit_code, _, stderr = self.RunBazel(
+      ['mod', 'graph', '--extension_info=all'],
+      rstrip=True, allow_failure=True,
+    )
+    self.AssertNotExitCode(exit_code, 0, stderr)
+    self.assertIn('\t\tfail("ext failed")', stderr)
+    self.assertIn('Error in fail: ext failed', stderr)
+
+  def testGraphWithFailingExtensionsKeepGoing(self):
+    # Force ext2 to fail.
+    with open(self.Path('MODULE.bazel'), 'a+') as f:
+      f.write("ext2.dep(name = 'repo2')\n")
+      f.write("ext2.dep(name = 'fail')\n")
+
+    _, stdout, stderr = self.RunBazel(
+      ['mod', 'graph', '--extension_info=all', '--keep_going'], rstrip=True
+    )
+    self.assertIn('\t\tfail("ext failed")', stderr)
+    self.assertIn('Error in fail: ext failed', stderr)
+    self.assertListEqual(
+      stdout,
```

Comment:
hmm... I guess I really don't understand how `--keep_going` interacts with Skyframe. On ModCommand.java:222, I'd expected the "has error" bit to be true for the evaluation, but apparently it just keeps going (heh) without considering that there's an error? Maybe it only considers there to be an error if the requested set of SkyKeys has errors?

Could you quickly check something -- is the error code on line 264 actually nonzero? If so, then I could still maybe make sense of the logic here.

Reply:
I checked that it's zero (the call above also doesn't pass `allow_failure=True`). 

Our observations are indeed consistent with Skyframe only setting the error bit if a top-level key evaluation fails. I agree that the exit code should be non-zero if any failing extension has been encountered, but I don't know how to pull that off. If we figure it out, we may be able to just get rid of the injected value.

I updated the test to verify exit codes (hence it fails).

Reply:
OK, I double-checked with @michaeledgar. The "keep going" setting makes it so that errors in transitive dep (Skyframe) nodes _don't_ bubble up to the surface, whereas in "no keep going" mode, as soon as a transitive dep node errors, its requesting node is also marked as erroring (even if it was try/catching, for example). This in turn means that the whole evaluation does not result in an error in "keep going" mode, as the evaluation roots are not erroring. And because we're always using "keep going" effectively (due to usage of `prepareAndGet`), the `mod` command has always had this behavior.

We should structure this in a way that the nodes whose deps are erroring _store_ the dep errors in its SkyValue (in this case, BazelModuleInspectorFunction should store extension eval errors in BazelModuleInspectorValue), and then later report those errors (if present) by checking the SkyValue (instead of just relying on the error status of the Skyframe eval result). Similar to how PackageFunction stores most errors in the `Package` object itself, instead of throwing an exception. (TIL more about Skyframe...)

As an aside, Mike also pointed out that we should not alter the behavior of SkyFunctions based on the value of `--keep_going`; any SkyFunction that _can_ keep going should _always_ "keep going". If `--keep_going` is false, Skyframe will not restart that SkyFunction anyway. That would also remove the `BAZEL_MOD_KEEP_GOING` precomputed value, which is nice.

Reply:
Thanks, this explanation was very helpful. I made the required changes.